### PR TITLE
Refactor and improve handling of invalid objects

### LINKED ISSUE #157: Decide how to deal with invalid objects
### Invalid attributes

There are many possibilities of invalid attributes of objects:
- Empty/invalid strings (for example forbidden characters in library element names)
- Negative numbers for line widths, drill diameters, pad sizes, and many more
- Invalid UUIDs, version numbers, ...
- And many more...

### Invalid combination of attributes

In addition, there are also *combinations* of attributes which I consider as (more or less) invalid:
- Footprints without any visible items (pads, polygons, ...)
- Components without any symbols
- Components which have multiple signals with the same UUID and/or name
- Pin-Signal-Mappings and Pad-Signal-Mappings which contain non-existing pins, pads or signals
- And many more...

Of course some of these invalid states can be avoided by validating new values before applying them, but some things are not really avoidable. For example when creating a new component, it won't have assigned any symbols yet, so there is a component without any symbols. Same for footprints, they are just empty right after creation.

Some things would be possible to avoid, but quite hard to implement. For example a component holds a list of its signals, where all signals must have unique names. But signals do not know each other and thus their method `setName(QString newName)` can't check whether there is already a signal with the same name.

These things are currently validated while saving the elements to their XML file. This was easy to implement, but it's quite annoying for users because they notice the errors only when trying to save the element. And even more annoying, it's currently impossible to save an invalid element, so if you want to stop working on a library element, you either need to fix all errors or discard all changes before you're able to close the library editor.

### Invalid attributes

I think it's important to avoid all smaller issues (negative numbers, invalid names, ...) as soon as possible:
1. Whenever possible, initialize attributes with valid, reasonable default values.
2. If 1. is not possible (for example there isn't a reasonable, valid default name for a newly created library element), these attributes must be passed to the constructor of the object to create. The constructor can then validate them and throw an exception if something is invalid.
3. All affected setter methods (for example `setLineWidth(int width)`) must not accept invalid values. They must throw an exception instead, with an expressive message (e.g. "The line width cannot be negative").

### Invalid combination of attributes

As many of these issues cannot be avoided with reasonable effort, maybe it's best to just accept them ;) So saving an invalid object (like a component which contains multiple signals with the same name) is possible. But there should be a way to validate such an element, for example add methods like `bool isValid()` and `QStringList getErrors()`, so the UI is able to display all issues of an invalid element.

@rnestler @dbrgn @0xB767B Any other ideas?

### LINKED ISSUE #288: Improve error messages for invalid library elements
##### LibrePCB VERSION

Commit d7d9096ec

##### OS / ENVIRONMENT

Arch Linux

##### SUMMARY

Please tell me what I did wrong :slightly_smiling_face: 

screenshot

### PR DESCRIPTION
**Pull request by ubruhin**
- Use `optional<T>` from https://github.com/TartanLlama/optional to represent nullable objects:
  - Use `optional<Uuid>` instead of nullable `Uuid`
    - Small change in the file format: Null UUIDs are now stored as `none` instead of `null`
  - Use `optional<Version>` instead of nullable `Version`
- Use `constrained_type<T>` from https://github.com/foonathan/type_safe to create constrained types:
  - Use constrained `Length` types `UnsignedLength` and `PositiveLength`
  - Use constrained `Ratio` type `UnsignedRatio`
  - Use constrained `QString` type `AttributeKey`
  - Use constrained `QString` type `GraphicsLayerName`
  - Use constrained `QString` type `ElementName`
  - Use constrained `QString` type `ComponentPrefix`
  - Use constrained `QString` type `ComponentSymbolVariantItemSuffix`
  - Use constrained `QString` type `CircuitIdentifier`
- `BoardDesignRules`: Disallow invalid design rules
- Generally relax some requirements on elements validity

The optional and constrained types are very useful to avoid invalid objects in many situations. Many invalid values are now rejected right in the user input forms and thus cannot propagate into the lower level classes.

Many elements have now pretty strong restrictions regarding the allowed characters. Especially symbol pins, package pads, component signals, component names and net signals must now follow the regex `^[-a-zA-Z0-9_+/!?@#$]{1,32}$`, otherwise they are rejected. Maybe we can relax these restrictions some time, but better start with strong restrictions and relax them later, the other way around would be much more cumbersome.

Fixes #157, Fixes #288

### CODE REVIEW

**Comment by dbrgn** [comment]
> Use optional<T> from https://github.com/TartanLlama/optional

Woah, that looks awesome :) Especially the map / and_then / etc stuff.

**Comment by ubruhin** [comment]
> Woah, that looks awesome :) Especially the map / and_then / etc stuff.

Yep ;)

I expect nobody wants to review this PR? :grin:

File: libs/librepcb/common/application.cpp [hunk 1]
```diff
@@ -37,35 +37,32 @@ namespace librepcb {
...
-               "Please correct this in the file 'librepcbcommon.pro'.");
+    // check "git describe" version
+    if (mGitVersion.isEmpty()) {
+        qWarning() << "Git revision not compiled into the executable!";
+    }
+
+    // check file format version
+    if (!mFileFormatVersion.isPrefixOf(mAppVersion)) {
+        qFatal("The file format version is not a prefix of the application version!");
+    }
```
Comment (dbrgn) [inline libs/librepcb/common/application.cpp]: Why do you pipe a string into qWarning, but call qFatal with a parameter?

Comment (ubruhin) [reply libs/librepcb/common/application.cpp]: Because qFatal does not provide the << operator.

**Pull request merged**